lib/pipeline.js

```diff
@@ -1,29 +1,42 @@
+const createPipelineStream = require('./createPipelineStream')
 const ns = require('./namespaces')
-const once = require('lodash/once')
-const Readable = require('readable-stream').Readable
 const Logger = require('./logger')
 
-class Pipeline extends Readable {
+class Pipeline {
+  /**
+   * Creates a new Pipeline stream
+   * The actual object returned by the constructor is the stream and not the pipeline instance
+   * @param node
+   * @param basePath
+   * @param context
+   * @param objectMode
+   * @param variables
+   * @param loaderRegistry
+   * @returns {*}
+   */
   constructor (node, { basePath = process.cwd(), context = {}, objectMode, variables = new Map(), loaderRegistry } = {}) {
-    super({ objectMode })
-
     this.node = node
     this.basePath = basePath
     this.variables = variables
+    this.loaderRegistry = loaderRegistry
+
+    this.readableObjectMode = Boolean(this.node.has(ns.rdf.type, ns.p.ReadableObjectMode).term)
+    this.readable = Boolean(this.node.has(ns.rdf.type, ns.p.Readable).term) || this.readableObjectMode
+    this.writableObjectMode = Boolean(this.node.has(ns.rdf.type, ns.p.WritableObjectMode).term)
+    this.writable = Boolean(this.node.has(ns.rdf.type, ns.p.Writable).term) || this.writableObjectMode
 
     this.context = { ...context }
     this.context.basePath = this.basePath
-    this.context.pipeline = this
-    this.loaderRegistry = loaderRegistry
-
     this.context.log = new Logger(this.node, { master: context.log })
 
-    this.init = once(() => this._init().catch(err => this.emit('error', err)))
-    this.initVariables = once(async () => {
-      const parsedVariables = await this.parseVariables()
-      this.variables = new Map([...parsedVariables, ...this.variables])
-      this.context.variables = this.variables
-    })
+    this.stream = createPipelineStream(this)
+    this.context.pipeline = this.stream
```

Comment:
Hm, this is suspicious. Before `context.pipeline` was assigned the pipeline itself and now it's a stream?

Reply:
I understand and I think it deserves a comment, but it's right. Before `context.pipeline` was assigned with `this`, cause there was only one pipeline object which represented the inner and outer pipeline. Now the stream is the outer/public interface. As the context should be used by other packages, the public interface should be assigned.

Reply:
Ok, makes sense. Should we still call this `context.pipeline`?